source/utils/algorithms.hpp

```diff
@@ -8,7 +8,7 @@
 namespace sim {
 
 using RoutingTable =
-    std::unordered_map<std::shared_ptr<IRoutingDevice>, std::unordered_map<std::shared_ptr<ILink>, int>>;
+    std::unordered_map<Id, std::unordered_map<std::shared_ptr<ILink>, int>>;
```

Comment:
Looks like a bug and might be done in new PR, but also might be done here. I think routing table should store `weak_ptr` to links:
```suggestion
    std::unordered_map<Id, std::unordered_map<std::weak_ptr<ILink>, int>>;
```